core/corerepo/gc.go

```diff
@@ -14,9 +14,27 @@ type KeyRemoved struct {
 	Key u.Key
 }
 
-func GarbageCollectBlockstore(n *core.IpfsNode, ctx context.Context) (<-chan *KeyRemoved, error) {
+func GarbageCollect(n *core.IpfsNode, ctx context.Context) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel() // in case error occurs during operation
+	keychan, err := n.Blockstore.AllKeysChan(ctx)
+	if err != nil {
+		return err
+	}
+	for k := range keychan { // rely on AllKeysChan to close chan
+		if !n.Pinning.IsPinned(k) {
+			err := n.Blockstore.DeleteBlock(k)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
```

Comment:
The meat of the operations is limited to 3 statements. The rest is boilerplate async. It seems that sharing more code would make the synchronous version more complex and the asynchronous one less direct.

```Go
keychan, err := n.Blockstore.AllKeysChan(ctx)
if !n.Pinning.IsPinned(k) {
err := n.Blockstore.DeleteBlock(k)
```